### PR TITLE
Implemented face detection filter

### LINKED ISSUE #14: Implement Face Detection filter
It would be nice to have face detection filter. Face detection could be made with Haar Cascades (https://docs.opencv.org/3.4.3/d7/d8b/tutorial_py_face_detection.html). 

Filter must be ROS nodelet and has to be located in https://github.com/ugnelis/ros_cameras_controller/tree/master/image_processing_filters. In this subproject there is a Threshold example.

### PR DESCRIPTION
**Pull request by rummanwaqar**
Closes #14 
Hacktoberfest :+1:

### CODE REVIEW

File: image_processing_filters/include/image_processing_filters/FaceDetectionFilterNodelet.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+     * @param message Image message.
+     */
+    void imageCallback(const sensor_msgs::ImageConstPtr& message);
+
+private:
+    std::shared_ptr<image_transport::ImageTransport> imageTransport_;   // Image receiver.
+    image_transport::Publisher imagePublisher_;                         // Image publisher.
+    image_transport::Subscriber imageSubscriber_;                       // Image subscriber.
+
+    std::string face_cascade_string_;
```
Comment (ugnelis) [inline image_processing_filters/include/image_processing_filters/FaceDetectionFilterNodelet.h]: Please use camel case.

File: image_processing_filters/include/image_processing_filters/FaceDetectionFilterNodelet.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    cv::CascadeClassifier face_cascade_;
```
Comment (ugnelis) [inline image_processing_filters/include/image_processing_filters/FaceDetectionFilterNodelet.h]: Please use camel case.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+PLUGINLIB_EXPORT_CLASS(image_processing_filters::FaceDetectionFilterNodelet, nodelet::Nodelet)
+
+
+namespace image_processing_filters {
+
+  FaceDetectionFilterNodelet::FaceDetectionFilterNodelet()
+        : imageTransport_(),
+          imagePublisher_(),
+          imageSubscriber_(),
+          face_cascade_string_(){
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: I missing space after `()`

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
+#include "image_processing_filters/FaceDetectionFilterNodelet.h"
+
+#include <pluginlib/class_list_macros.h>
+
+PLUGINLIB_EXPORT_CLASS(image_processing_filters::FaceDetectionFilterNodelet, nodelet::Nodelet)
+
+
+namespace image_processing_filters {
+
+  FaceDetectionFilterNodelet::FaceDetectionFilterNodelet()
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: Please keep consistency: Either after namespace use spaces before every functions or don't use any.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+          imagePublisher_(),
+          imageSubscriber_(),
+          face_cascade_string_(){
+  }
+
+  FaceDetectionFilterNodelet::~FaceDetectionFilterNodelet() {
+
+}
+
+void FaceDetectionFilterNodelet::onInit() {
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp:21]: Please keep consistency: Either after namespace use spaces before every functions or don't use any.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    ros::NodeHandle &nodeHandle = getNodeHandle();
+    ros::NodeHandle &privateNodeHandle = getPrivateNodeHandle();
+
+    imageTransport_ = std::make_shared<image_transport::ImageTransport>(nodeHandle);
+    imagePublisher_ = imageTransport_->advertise("face_detect", 1);
+    imageSubscriber_ = imageTransport_->subscribe("image_raw", 1, &FaceDetectionFilterNodelet::imageCallback, this);
+
+    privateNodeHandle.param<std::string>("classifier_file", face_cascade_string_, "haarcascade_frontalface_alt.xml");
+
+    if( !face_cascade_.load( face_cascade_string_ ) ) {
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: Please, don't use old school spacing. It should be: `if (!face_cascade_.load(face_cascade_string_)) {`.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        ROS_ERROR("cv_bridge exception: %s", e.what());
+    }
+
+    cv::Mat receivedImage = cvImagePtr->image;
+    cv::Mat greyImage;
+    cv::cvtColor(receivedImage, greyImage, CV_RGB2GRAY);
+
+    //-- Detect faces
+    std::vector<cv::Rect> faces;
+    face_cascade_.detectMultiScale( greyImage, faces, 1.1, 2, 0|CV_HAAR_SCALE_IMAGE, cv::Size(30, 30) );
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: Old school spacing :(

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    for( size_t i = 0; i < faces.size(); i++ )
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: Bad spacing.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    {
+      cv::Point center( faces[i].x + faces[i].width*0.5, faces[i].y + faces[i].height*0.5 );
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: It should be `cv::Point center(faces[i].x + faces[i].width * 0.5, faces[i].y + faces[i].height * 0.5);`

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      cv::ellipse( receivedImage, center,
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: Spacing.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+                   cv::Size( faces[i].width*0.5, faces[i].height*0.5),
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: Spacing.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+                   0, 0, 360, cv::Scalar( 255, 0, 255 ), 4, 8, 0 );
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: Spacing.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 13]
```diff
@@ -0,0 +1,70 @@
+#include "image_processing_filters/FaceDetectionFilterNodelet.h"
+
+#include <pluginlib/class_list_macros.h>
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp:1]: This include should be first because it is not local include.

**Review by ugnelis — CHANGES_REQUESTED** [review]
Overall, nice job, but, please, fix code style.

**Comment by rummanwaqar** [comment]
@ugnelis formatting is fixed.

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 14]
```diff
@@ -48,14 +48,14 @@ void FaceDetectionFilterNodelet::imageCallback(const sensor_msgs::ImageConstPtr
 
     //-- Detect faces
     std::vector<cv::Rect> faces;
-    face_cascade_.detectMultiScale( greyImage, faces, 1.1, 2, 0|CV_HAAR_SCALE_IMAGE, cv::Size(30, 30) );
+    faceCascade_.detectMultiScale(greyImage, faces, 1.1, 2, 0|CV_HAAR_SCALE_IMAGE, cv::Size(30, 30));
 
     for( size_t i = 0; i < faces.size(); i++ )
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: Should be `for (size_t i = 0; i < faces.size(); i++) {`

File: image_processing_filters/src/FaceDetectionFilterNodelet.cpp [hunk 15]
```diff
@@ -26,10 +26,10 @@ void FaceDetectionFilterNodelet::onInit() {
     imagePublisher_ = imageTransport_->advertise("face_detect", 1);
     imageSubscriber_ = imageTransport_->subscribe("image_raw", 1, &FaceDetectionFilterNodelet::imageCallback, this);
 
-    privateNodeHandle.param<std::string>("classifier_file", face_cascade_string_, "haarcascade_frontalface_alt.xml");
+    privateNodeHandle.param<std::string>("classifier_file", faceCascadeString_, "haarcascade_frontalface_alt.xml");
 
-    if( !face_cascade_.load( face_cascade_string_ ) ) {
-      ROS_ERROR_STREAM("error loading cascade classifier: " << face_cascade_string_);
+    if(!faceCascade_.load(faceCascadeString_)) {
```
Comment (ugnelis) [inline image_processing_filters/src/FaceDetectionFilterNodelet.cpp]: It should be `if<space>(`.

**Review by ugnelis — CHANGES_REQUESTED** [review]
Some fixes should be made.

**Review by ugnelis — APPROVED** [review]
Good job :)

**Pull request merged**